src/OrchardCore/OrchardCore.Users.Core/Indexes/UserRoleIndex.cs

```diff
@@ -0,0 +1,37 @@
+using System.Collections.Generic;
+using System.Linq;
+using OrchardCore.Users.Models;
+using YesSql.Indexes;
+
+namespace OrchardCore.Users.Indexes;
+
+public class UserRoleIndex : MapIndex
+{
+    public string UserId { get; set; }
+
+    public string Role { get; set; }
+}
+
+public class UserRoleIndexProvider : IndexProvider<User>
```

Comment:
@sebastienros the `UserByRoleNameIndex` provided you with how many users are in each role. The new `UserRoleIndex` will provide user to role index so we know what roles each user has. This will allow us to know which user can list which roles.

Reply:
`Count` is a by-product of the reduce index. Ignore that. Here is how it is used in a query: https://github.com/OrchardCMS/OrchardCore/blob/70619ed8020a61efc946910c64f19421fb97d911/src/OrchardCore/OrchardCore.Users.Core/Services/UserStore.cs#L460

Isn't it what you want to do? List all users in specific roles?

Reply:
I am still trying to understand how that would work but it did! I removed all the index

Reply:
Yes and just for info if you had kept your new index, better to use a migration step than a new Migrations create() because a deferred task was used to update users, this to prevent to run this deferred task on each new installation.

Reply:
@jtkech are you saying place the update code in the `Create()` method instead of calling a differed task from the Migration? Because last change I did is calling a deferred task to update the users in a batch.

Reply:
Just for info for this question I missed, what I meant from memory.

So, when we add migration steps the rule is to update the Create() method so that on a new installation this method does all what is needed, and at the end of Create() return the very last version number so that these migration steps are skipped on a new installation.

For example in the Create() method we may change table definitions but we don't add something and then remove it, we just don't define it anymore, idem for any custom code that is only intended to be run on an existing installation.

That said, when we create a very new `DataMigration` class, its Create() method will be called on any new installation, so if the code is only intended to migrate existing installations, this is not the right place, unless we can add a check to see if it is already migrated or a new installation.

So if possible better to use a migration step as described above with an existing `DataMigration`, for example if one exists in a given module, otherwise we have no choice but still need the check to simulate a migration step.

Hmm, maybe worth to do something oob, a kind of global migrations version that would be persisted on any setup, so that on any startup if it is lower (or doesn't exist) we could run some code or not, e.g. in a new Migrations Create() triggered by the AutomaticDataMigrations.

Reply:
@jtkech thank for that info. That migration was removed from this PR. We are no longer adding new index or running migration